### PR TITLE
[bevy_ui/layout] Handle ui heirchy change edge cases

### PR DESCRIPTION
**Pull request by StrikeForceZero**
- [ ] I need to open an issue that documents the problem this is solving

Blocked on:
- [x] https://github.com/bevyengine/bevy/pull/12803
- [ ] ~~https://github.com/bevyengine/bevy/pull/12804~~ https://github.com/bevyengine/bevy/pull/16362

---

# Objective

- Handle UI Nodes being moved to a new parent `Changed<Parent>`
- Handle UI Nodes who lost their parent `RemovedComponent<Parent>`

## Solution

- Adds query to the `layout_system` to handle the specified edge cases

## Testing

- Did you test these changes? If so, how?
  - Added Tests
- Are there any parts that need more testing?
  - Identifying any potential regressions or unknown side effects
- How can other people (reviewers) test your changes? Is there anything specific they need to know?
  - Testing an example project that emulates a drag-and-drop-like behavior for UI node relationships. 
  - Think of alternative edge cases that fit with the scope of this PR.
- If relevant, what platforms did you test these changes on, and are there any important ones you can't test?
  - Tested on Linux

---

## Changelog

- Added handling of UI nodes changing parents
- Added handling of UI nodes removing parents

### CODE REVIEW

File: crates/bevy_ui/src/layout/mod.rs [hunk 1]
```diff
@@ -158,6 +160,82 @@ pub fn ui_layout_system(
         }
     }
 
+    for new_root_node_entity in removed_parents.read() {
+        // TODO: most of this should be moved into a function on `UiSurface` but the query requirements might feel out of place
```
Comment (StrikeForceZero) [inline crates/bevy_ui/src/layout/mod.rs:164]: I was planning on moving this logic into an associated function on `UiSurface`, but passing the queries seems out of scope. Open to suggestions.

**Comment by StrikeForceZero** [comment]
superseded by https://github.com/bevyengine/bevy/pull/16383
